### PR TITLE
Add boost.test support to ApprovalTests.cpp

### PR DESCRIPTION
**Pull request by claremacrae**
Specify library name and version:  **ApprovalTests.cpp - adding Boost.Test support**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

**Changes made:**

1. Fix a logic error in the existing code, that the doctest test was not run, due to accidental `elif` instead of `if`
2. Added some print statements to the test running code, to confirm that tests are actually all being run (as some of the test frameworks are very quiet)
3. Add a 4th option to the ApprovalTests.cpp recipe, `with_boosttest` - with test added. This actually runs an approval test, to confirm that the framework is set up correctly.

**Comments/Questions**

1. `with_boosttest` or `with_boost_unit_test_framework`?  
We saw that the boost test framework is named `boost_unit_test_framework` in some of the Conan output - would you like us to rename the option `with_boosttest` to `with_boost_unit_test_framework`

### CODE REVIEW

**Comment by claremacrae** [comment]
Apologies for the build failure... The message I got was:

Some configurations of 'approvaltests.cpp/8.0.0' failed in build 1

The new feature will only work with ApprovalTests v.8.6.0 and above, so we will have to work out how to constrain the new feature...

We are next working on together on Monday, so will pick this up then.

**Comment by dheater** [comment]
The build error is reproducible locally by running `conan create . approvaltests.cpp/8.0.0@`. I think it comes down to the fact that `boosttest` is not a valid option for earlier versions of ApprovalTests.cpp, but all versions are sharing the same Conan recipe and tests.

I'm not sure if there is a way to support this without creating a second recipe in a second folder. See "N versions => M recipes" in https://github.com/conan-io/conan-center-index/wiki

**Comment by danimtb** [comment]
You can delete the "boosttest" option conditionally to the version you are using in the configure() method. See https://github.com/conan-io/conan-center-index/blob/master/recipes/openssl/all/conanfile.py#L145L149

**Comment by claremacrae** [comment]
> You can delete the "boosttest" option conditionally to the version you are using in the configure() method. See https://github.com/conan-io/conan-center-index/blob/master/recipes/openssl/all/conanfile.py#L145L149

Thanks - yes, that bit is working fine for me.

What I'm struggling with is modifying the builds in `test_package` to prevent it trying to build the Boost.Test-based tests...

How can I find the approvaltests.cpp version number or its supported options in [recipes/approvaltests.cpp/all/conanfile.py](https://github.com/conan-io/conan-center-index/blob/master/recipes/approvaltests.cpp/all/conanfile.py) please?

**Comment by claremacrae** [comment]
According to Hipony on cpp Slack:

> The problem is that `configure` is run _before_ `requirements` it seems

In the tests, I was trying to remove the boost option in `configure()` if using ApprovalTests earlier than `8.6.0` - and this line was failing with a `KeyError`:

```python
    print(self.deps_cpp_info['approvaltests.cpp'].version)
```

It turns out to be harmless to leave an unsupported value in the tests:

```python
class TestPackageConan(ConanFile):
    settings = "os", "compiler", "arch", "build_type"
    generators = "cmake"
    default_options = {
        "approvaltests.cpp:with_boosttest": True,
        "approvaltests.cpp:with_catch2": True,
        "approvaltests.cpp:with_gtest": True,
        "approvaltests.cpp:with_doctest": True
    }
```

File: recipes/approvaltests.cpp/all/conanfile.py [hunk 1]
```diff
@@ -12,22 +13,31 @@ class ApprovalTestsCppConan(ConanFile):
...
         "with_catch2": False,
         "with_gtest": False,
         "with_doctest": False
     }
     no_copy_source = True
 
+    def configure(self):
+        if not self._boost_test_supported():
+            del self.options.with_boosttest
+            self.default_options.__delitem__("with_boosttest")
```
Comment (uilianries) [inline recipes/approvaltests.cpp/all/conanfile.py:32]:
```suggestion
```

Only `del self.options` is enough

File: recipes/approvaltests.cpp/all/conanfile.py [hunk 2]
```diff
@@ -12,22 +13,31 @@ class ApprovalTestsCppConan(ConanFile):
...
+
     @property
     def _header_file(self):
         return "ApprovalTests.hpp"
 
     def requirements(self):
+        if self._boost_test_supported() and self.options.with_boosttest:
```
Comment (uilianries) [inline recipes/approvaltests.cpp/all/conanfile.py:39]:
```suggestion
        if self.options.get_safe("with_boosttest"):
```

As with_boosttest can be deleted, the method `get_safe` returns None when it's not available, or True/False when available.

**Comment by claremacrae** [comment]
Hi @uilianries - thanks for the code review - very helpful! :-)

I've accepted the changes, and it all works for me with both 8.5.0 and 8.6.0, so back to you!

File: recipes/approvaltests.cpp/all/test_package/.gitignore [hunk 3]
```diff
@@ -0,0 +1 @@
+*.received.txt
```
Comment (jgsogo) [inline recipes/approvaltests.cpp/all/test_package/.gitignore:1]: Is this file needed in this package?

Comment (dheater) [reply recipes/approvaltests.cpp/all/test_package/.gitignore:1]: `*.received.txt` files are generated as a side effect of running the test. I don't know if that qualifies it as **needed**, but it is a convenience to help prevent accidental inclusion of these test outputs in a commit.

Comment (jgsogo) [reply recipes/approvaltests.cpp/all/test_package/.gitignore:1]: Ok, it the automatic checks don't complain... I'm nobody to say anything 😄
